### DISCUSSION TITLE
Minimum number of parameters for params

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5477] [Language Ideas]
Can we introduce an attribute or a keyword which stipulates that `params` receives a minimum number of items?
```cs
void Foo([MinimumCount(1)] params string[] list) { ... }

Foo(); // compilation error

string[] arr = Array.Empty<string>();
Foo(arr); // compiles, but runtime exception: ArgumentInsufficientCountException

Foo("abc"); // compiles and runs
Foo("abc", "def"); // compiles and runs
```
Motivation: this is useful for methods like `MakeGenericType` and all the other similar methods, as well as custom code which utilizes these reflection methods; it's impossible to make a generic type/method without at least one type argument.

Workaround: 
```cs
void Foo(string a, params string[] extra)
{
  var list = new string[extra.Count+1];
  list[0]=a;
  Array.Copy(extra, list, 1);
}
```
You can see how the workaround has an ugly signature and overhead of allocating another array internally (assuming it's usually needed to have the complete list).

**Comment by jnm2** [comment]
This can be implemented as an analyzer package that defines the attribute and enforces the usage.

**Reply by TahirAhmadov** [reply]
Partially. Analyzer cannot emit runtime code to validate the parameter for length; and given how many methods in BCL need this, it makes sense to add it to the framework.

**Reply by HaloFour** [reply]
> and given how many methods in BCL need this, it makes sense to add it to the framework.

How many BCL methods use `params` and have a requirement to pass a minimum number of arguments?  I can't recall encountering one and I would argue that such is a terrible design choice.

**Reply by jnm2** [reply]
So it would be attribute plus manual `if (p.Length == 0) throw new ArgumentException(...);`, which the analyzer can also require you to add so you don't forget it, and even provide a code fix to insert it.

Then the value of a hypothetical language feature would be to save you from having to use the light bulb to insert the check or having to read it later.

**Reply by TahirAhmadov** [reply]
> > and given how many methods in BCL need this, it makes sense to add it to the framework.
> 
> How many BCL methods use `params` and have a requirement to pass a minimum number of arguments? I can't recall encountering one and I would argue that such is a terrible design choice.

It's in the OP - all the reflection methods like `MakeGenericType`.

**Reply by HaloFour** [reply]
Those are advanced methods that need to be used precisely to work anyway.  It's not that they require at least one argument, it's that they require the exact number of arguments to match the expected open generic type, which isn't known at compile-time anyway.

**Reply by TahirAhmadov** [reply]
> So it would be attribute plus manual `if (p.Length == 0) throw new ArgumentException(...);`, which the analyzer can also require you to add so you don't forget it, and even provide a code fix to insert it.
> 
> Then the value of a hypothetical language feature would be to save you from having to use the light bulb to insert the check or having to read it later.

Yes, I could do all that. Still, it isn't propagated to the BCL, and it's a custom thing I can enforce on my codebase only - every team will have to add something like this, which is unlikely to happen. The benefit of introducing this into the framework and language is that it's a validation which is easy to use out of the box - many more teams will start using it. I mean, I'm the author and I'm not sure I value "checking myself" enough to devote the time to developing/testing/applying the analyzer. (I may do it - just a question of priorities.)

**Reply by TahirAhmadov** [reply]
> Those are advanced methods that need to be used precisely to work anyway. It's not that they require at least one argument, it's that they require the exact number of arguments to match the expected open generic type, which isn't known at compile-time anyway.

The precise # of types is unknown at compile time, but we do know it's >=1 at compile time. Bugs can be avoided if there is a typo and no types were passed in to those methods. Yes it won't prevent somebody passing in too many arguments, etc., but it's better than nothing.

**Reply by jnm2** [reply]
As far as having the BCL use this, the BCL methods are already throwing argument exceptions where they need to, and catching at compile time that zero types were passed to MakeGenericType/Method is something that seems to better fall under the BCL's existing push to add BCL usage analyzers that they consider worth it.

And even there, I'm not sure the BCL would consider MakeGenericType/Method worth it for the reason @HaloFour already mentioned, though there's certainly no harm in starting a discussion in the runtime repo.

**Reply by jnm2** [reply]
> but it's better than nothing.

"Better than nothing" isn't the benchmark though; it has to be better than paying the cost of adding the analyzer or creating a new language feature, and I'm skeptical of both. I'm not sure I've heard of this mistake happening before in ten years. It's also not a particularly difficult thing to react to when you get the runtime exception in the BCL.

**Reply by TahirAhmadov** [reply]
Also, methods such as `string.Format` don't have a strict requirement for at least one parameter, but there is definitely a soft requirement. There can be a parameter on the attribute; if true, treat as error, and if false, treat as warning. So all the `Format` methods can be have the soft minimum; it'll remind people to get rid of the `Format` if  they don't need to use it.

**Reply by jnm2** [reply]
`string.Format` is an example of where a custom-tailored analyzer that can see the `{0}` and `{1}` in string literals would be better than an approach based on `[MinimumCount(1, error: false)]` or similar.

**Reply by TahirAhmadov** [reply]
I do agree this isn't high priority; just wanted to start a conversation about it. I think it has some value, but I'm not insisting it has to be included in the next version or anything like that.

**Reply by TahirAhmadov** [reply]
> `string.Format` is an example of where a custom-tailored analyzer that can see the `{0}` and `{1}` in string literals would be better than an approach based on `[MinimumCount(1, error: false)]` or similar.

I was thinking more `string.Format("Hello, dear user!")`. At some point in the past, it was `string.Format("Hello, {0}!", name);`. The requirement to show the name was dropped, but an unnecessary call to `string.Format` remains.
Or even more likely, `writer.WriteFormat("Hello, user!");` because people just copy and paste without thinking; it'll remind them to use `writer.Write("Hello, user!");` which has less overhead.
Again, not a huge issue; more of a nice to have.

**Reply by jnm2** [reply]
@TahirAhmadov That's exactly what I'm talking about though; the same analyzer that should tell you about your mistake if you do `string.Format("{0}")` could also tell you about `string.Format("x")`. I'm not seeing `[MinimumCount(1, error: false)]` as necessarily the better path to that end.

**Reply by TahirAhmadov** [reply]
> @TahirAhmadov That's exactly what I'm talking about though; the same analyzer that should tell you about your mistake if you do `string.Format("{0}")` could also tell you about `string.Format("x")`.

True - however now you can see that the 2 options are 1) the framework and language are modified a little to support this and every team is nudged to write cleaner code; or 2) it's something every team CAN implement as custom analyzers, which will never happen. I'm not so much asking for myself - for me it's more to improve the overall quality.

**Reply by jnm2** [reply]
Just to clarify; it's not about priority so much as the fact that the existence of the BCL analyzer or the existence of the language feature themselves are a perpetual cost. Even if there was no other feature going into the BCL or language to compete in terms of priority, this cost could mean that this proposal would not be desired.

**Reply by jnm2** [reply]
> True - however now you can see that the 2 options are 1) the framework and language are modified a little to support this and every team is nudged to write cleaner code; or 2) it's something every team CAN implement as custom analyzers, which will never happen. I'm not so much asking for myself - for me it's more to improve the overall quality.

So the BCL is addressed pretty well by BCL analyzers then, and community APIs are addressed pretty well (IMO) by a reusable community package providing `MinimumCountAttribute` + analyzer. Once written by one person, every team around the world can use it immediately. (If it was a language feature, you'd have to wait something like five years from now.)

**Reply by TahirAhmadov** [reply]
WRT `MakeGenericType`, I thought of this idea of adding a BCL attribute and a compiler error/warning because the ">=1" rule is so fundamental to those methods, it's basically part of the signature (albeit not "written"/"enforced" right now). I think of it like, `MakeGenericType(params object[] types)` is almost as bad compared to `MakeGenericType(params Type[] types)`, as the latter is compared to `MakeGenericType([MinimumCount(1, error: true)] params Type[] types)`.

**Reply by TahirAhmadov** [reply]
> > True - however now you can see that the 2 options are 1) the framework and language are modified a little to support this and every team is nudged to write cleaner code; or 2) it's something every team CAN implement as custom analyzers, which will never happen. I'm not so much asking for myself - for me it's more to improve the overall quality.
> 
> So the BCL is addressed pretty well by BCL analyzers then, and community APIs are addressed pretty well (IMO) by a reusable community package providing `MinimumCountAttribute` + analyzer. Once written by one person, every team around the world can use it immediately. (If it was a language feature, you'd have to wait something like five years from now.)

True - perhaps I should create that package. In the description, I will write "not using this package is SINFUL" - hopefully that has the desired effect. :)

**Reply by jnm2** [reply]
I don't think I can evaluate "seems fundamental" other than by considering empirically how much time and energy it would save people. How often does this happen and how costly is it to fix when it does happen? Do the victims tend to wish they weren't C# developers at this point or do they tend to be amused that they missed this parameter?

**Reply by TahirAhmadov** [reply]
> I don't think I can evaluate "seems fundamental" other than by considering empirically how much time and energy it would save people. How often does this happen and how costly is it to fix when it does happen?

The "fundamental" aspect is more of a logical question in my opinion, not a practical one. If the type was `params object[] types` hypothetically, but that method was almost never called with anything but `Type`s, it wouldn't make sense "practically" to change it to `params Type[] types`, but it still makes sense "logically". (Obviously it's a silly example to illustrate my point.) Again, I do hear your concern than the practicality is outweighed by the cost; I don't know enough about the actual BCL and Roslyn and compiler development costs to argue with that either way.

**Reply by jnm2** [reply]
I don't see `Type[]` being better than `object[]` as anything other than a practical issue, backed by quite some evidence. I'd have a hard time defining what "logical" would mean in the absence of the practical. Does it mean some particular person's intuition? Then you'd better hope everybody already has that intuition.

(If someone showed me evidence that writing more kinds of unit tests was more productive than strong typing, and I could reproduce this effect for myself, I'd switch eagerly.)

**Reply by TahirAhmadov** [reply]
> I don't see `Type[]` being better than `object[]` as anything other than a practical issue, backed by quite some evidence. I'd have a hard time defining what "logical" would mean in the absence of the practical. Does it mean some particular person's intuition? Then you'd better hope everybody already has that intuition.
> 
> (If someone showed me evidence that writing more kinds of unit tests was more productive than strong typing, and I could reproduce this effect for myself, I'd switch eagerly.)

Ultimately, it boils down to what works best, but some "theorems" are proven once and can be relied on going forward. We don't need to re-evaluate strong typing over and over - it's a waste of time - we know strong typing is better. There is a "mathematical" truth to it, IMO. It's like saying, 2+2=4 is a statement we have to evaluate for its practicality. Yes we _could_ do that, but then we'll spend all of our time questioning everything - not a very practical thing, you know.

**Reply by jnm2** [reply]
I disagree that the value of `MinimumCountAttribute` can be automatically equated with the value of existing strong typing, though, certainly not with the clarity of a mathematical expression. The tradeoffs involved in reading and writing this particular syntax are unique to this question (even if it's a new keyword). It wouldn't be a re-evaluation.

**Reply by TahirAhmadov** [reply]
> I disagree that the value of `MinimumCountAttribute` can be automatically equated with the value of existing strong typing, though, certainly not with the clarity of a mathematical expression. The tradeoffs involved in reading and writing this particular syntax are unique to this question (even if it's a new keyword). It wouldn't be a re-evaluation.

Oh yes, I never said it was equivalent. But it is similar in the sense that the min # (in the `MakeGenericType` case) is part of the signature, and so are strong types.

**Reply by Joe4evr** [reply]
> How many BCL methods use `params` and have a requirement to pass a minimum number of arguments? I can't recall encountering one and I would argue that such is a terrible design choice.

[The API committee reviewed such APIs for the BCL last month, and they decided to do the exact thing that OP lists as a "workaround".](https://github.com/dotnet/runtime/issues/54220#issuecomment-946994846) I'd regard that as a good argument to consider that to be the canonical solution to this problem.

**Comment by CyrusNajmabadi** [comment]
These are well known symbols. I feel like we don't even need an attribute. One could just write a basic analyzer here. That said, I'm skeptical of the value. How often does one call this method and pass no Params?  Fellas like it would be caught immediately.

**Reply by TahirAhmadov** [reply]
That's what the thread arrived at, too - a Nuget package with the attribute type and an analyzer.

**Comment by Richiban** [comment]
I have also needed this in the past (enough times to make it worth mentioning), at least a way of saying "You can pass in as many arguments as you like, but you must pass at least one". I have also implemented the _workaround_ of having a regular parameter for the first element and a `params` array for the rest and but this is very unsatisfactory for two reasons:

1. If the caller already has an array they now can't pass it straight to your method any more; they must split the array into head and tail and pass them separately which is wasteful
2. In the method implementation, if what you needed was actually to store an array of items, you've now got to recreate it yourself from the head and tail which is, again, wasteful

I'm not really sure what the correct answer is here, so maybe an analyzer really is the best approach since it's the least invasive to implement. Having said that, I could go on a slightly separate rant about how I so often wish there were types to represent "non-empty" collections in the BCL.
